### ISSUE TITLE
Mapping too slow with nested object - compilation granularity (Source w/o destination)

### ISSUE THREAD

**Issue by Mariana1980** [expressmapper/ExpressMapper#53]
Hello!!!
We are testing your mapping class and we have a problem. Our class has a complex object list (Unidad).
The first time mapping is very slow , apox. 2 minutes.  The second time is immediately. 
What can we do to improve the performance? 

WPF App:
   public partial class App : Application
    {
        private void OnStartup(object sender, StartupEventArgs e)
        {

            Mapper_Inspector_Punitiva mapper = new Mapper_Inspector_Punitiva();
            mapper.RegisterMapping();
            Mapper.Compile();
            Mapper.PrecompileCollection<List<pesada_punitiva>, List<Pesada>>(); // 2 minutes
            
        }
    }

Class:
    [Serializable]
    public class Pesada
    {      
        public long id_pesada { get; set; }   
 public List<Unidad> unidades_punitiva { get; set; }             
    }

        [Serializable]
    public class Unidad
    {
        public long id_pesada { get; set; }
        public int id_balanza { get; set; }
        public System.DateTime fecha_pesada { get; set; }
        public int id_unidad { get; set; }   

        public List<Grupo> grupos_punitiva { get; set; }
        public List<Infraccion> infracciones_punitiva { get; set; }
    }

   [Serializable]
    public class Grupo
    {
        public long id_pesada { get; set; }
        public System.DateTime fecha_pesada { get; set; }
        public int id_balanza { get; set; }
        public int id_unidad { get; set; }
        public int id_grupo { get; set; }
        public long? peso_maximo { get; set; }
        public int? peso_grupo { get; set; }
        public int? tolerancia_adicional { get; set; }
        public int? tipificacion_matricula_cantidad_neumaticos_dnt { get; set; }
        public string tipificacion_matricula_suspension_dnt { get; set; }
        public string tipificacion_matricula_tipo_neumaticos_dnt { get; set; }
        public List<Eje> ejes_punitiva { get; set; }
    }

    [Serializable]    
    public class Eje
    {
        
        public long id_pesada { get; set; }
        public System.DateTime fecha_pesada { get; set; }
        public int id_balanza { get; set; }
        public int id_unidad { get; set; }
        public int id_grupo { get; set; }
        public int id_eje { get; set; }
        public long? distancia_eje { get; set; }
        public int? peso_eje { get; set; }
        public int? tolerancia_adicional { get; set; }
        public int? tipificacion_matricula_cantidad_neumaticos_dnt { get; set; }
        public string tipificacion_matricula_suspension_dnt { get; set; }
        public string tipificacion_matricula_tipo_neumaticos_dnt { get; set; }
        public int? peso_maximo { get; set; }
        public List<InfoSensorPeso> info_sensor_peso_punitiva { get; set; }
    }

Mapping:

            Mapper.Register<pesada_punitiva, Pesada>();
            Mapper.Register<unidad_punitiva, Unidad>();
            Mapper.Register<grupo_punitiva, Grupo>();
            Mapper.Register<eje_punitiva, Eje>();

**Comment by anisimovyuriy** [comment]
Dear Mariana1980,

Thanks a lot for your input!
First of all I'd like to clarify some points:
1) You specified it happened on Startup_App on ```Mapper.PrecompileCollection<List<pesada_punitiva>, List<Pesada>>();``` is that right (trying to precompile collections)? Or does it happen when you try to really map ```List<pesada_punitiva>``` to ```List<Pesada>``` using Mapper.Map<List<pesada_punitiva>, List<Pesada>>(src)?
3) Do you use exactly the same mapping types in your client code that you've specified in registration phase ```List<pesada_punitiva>``` to ```List<Pesada>```
2) Can you send me a declaration for all of your Source (Domain level objects) and Destination (DTO). Like you did for Eje, Grupo, Unidad, Pesada, so I need declarations for pesada_punitiva, unidad_punitiva, grupo_punitiva, eje_punitiva in order to reproduce it.
4) I tried on my side with nested collections plus nested associations - precompilecollection works 3-6ms and mapping performed very fast as well - of course it depends which kind of volume you are trying to map.

Looking forward for your answers,

Thanks,
Yuriy

**Comment by anisimovyuriy** [comment]
Dear Mariana1980,

I've just created a test project to test performance. You can downloaded from <a href="https://www.dropbox.com/s/0a6z5v12d7d2l7b/TestMapperSpeed.zip?dl=0">here</a>

The results are pretty interesting! I ran the tests on 2 machines (Test project's been built with .NET 4.0 ).
The first one is native windows machine with VS2013:
![mappertest_vs2013_native_windows](https://cloud.githubusercontent.com/assets/3758138/11054912/28ad4294-8726-11e5-87b6-32792b692f33.PNG)

The second one is my VMWARE FUSION 8 Virtual machine on my MAC with VS2015:
<img width="912" alt="mappertest_vs2015_vm_windows" src="https://cloud.githubusercontent.com/assets/3758138/11054937/701c6ef2-8726-11e5-8691-f7589765ee91.png">

I've tried to debug whether the first time map uses ```Mapper.PrecompileCollection<List<Order>, List<OrderViewModel>>()``` from cache and yes I confirm that it was taken from cached the first and the second time in the test mapping has been taken from cache.

I can tell you for sure 100% this is how JIT optimization translates from MSIL to windows native code - so it depends on many factors but the code is executed the same way twice and it is pre-cached - so it completely depends on .Net JIT compiler.

Regarding your question I don't know exactly if you have the same context can you answer my previous question and let's see what I can do.

Thanks for your input,
Yuriy

**Comment by Mariana1980** [comment]
Hello Yuriy!!
Thanks for your soon answer!!!
We used precompile because the mapping was too slow (2 minutes)  in the line:
```
 List<Pesada> lstPesada_aux = Mapper.Map<List<pesada_punitiva>, List<Pesada>>(new List<pesada_punitiva>());
```
With Precompile the performance is the same (2 minutes) in the line 
```
Mapper.PrecompileCollection<List<pesada_punitiva>, List<Pesada>>();
```
but in this case the mapping works vary fast, after precompile. 
This is without loading data.

Now, we have your TestMapperSpeed project with our classes, but in this case the application dies after mapping the empty list. 

You can downloaded from [here] (https://www.dropbox.com/s/lqsubfuepyxrs0z/TestMapperSpeed.zip?dl=0)
We compile with .NET 4.5

Thanks for your time,
Mariana

**Comment by anisimovyuriy** [comment]
Dear Mariana,

Thanks a lot for your valuable input! It's very good test for Expressmapper having such huge amount nested associations and collections.
So I checked it on my side ```Mapper.PrecompileCollection<List<pesada_punitiva>, List<Pesada>>();``` works and it caches ```Func<List<pesada_punitiva>, List<Pesada>>``` so any upcoming mapping will use it from the cache. 
Here is the first result using Mapper.PrecompileCollection (using your solution, log has been added for precompile collection and mapping as separate logs):
![results](https://cloud.githubusercontent.com/assets/3758138/11084258/f4dabab4-87ed-11e5-94ea-1323702bc3c0.PNG)

Here is the second result without precompilation collection step (using your solution with commented Mapper.PrecompileCollection and mapping executing twice):
![resultswithoutprecompilation](https://cloud.githubusercontent.com/assets/3758138/11084290/77db122e-87ee-11e5-8e6c-dc2fe7327f2e.PNG)

So from that point if you don't use precompilation it will be compiled when you try mapping first time which is obvious and explained at <a href="http://expressmapper.org">http://expressmapper.org</a>.

Regarding the why it was stuck on your machine - maybe different hardware and software as I tried on both machines and it worked out for me.

What and why it takes so long time to precompile collection? :
There are 2 compilations are happening when you try to compile:
1) ```Mapper.Map<TSource, TDestination>(source)```
2) ```Mapper.Map<TSource, TDestination>(source, destination)```

The first one works in your case very fast as it is simple algorithm even with your number of classes involved (116 classes plus multiple usages inside).
The second one takes a lot time when it tries to compile to ```Func<>``` because there are a lot of code to compile in your case as we are tracking instances of objects and many other things.
You can take a look at real expression trees for your case for the first case 
<a href="https://www.dropbox.com/s/e7jmp7fvd4mh7z1/SourceCollectionPrecompilation.txt?dl=0">here</a>  file size 76KB and for the second case <a href="https://www.dropbox.com/s/3qtnfnitws4orvk/WithDestinationCollectionPrecompilation.txt?dl=0">here</a> file size 13MB. Which is completely fair with such complex mappings as yours and with our code to support to map to existing destination without creating extra instances of destination type.

If you don't need mapping to existing destination ```Mapper.Map<TSource, TDestination>(source, destination)``` I think I can help you out. I can release version with extended functionality to compile only source mappings that improve your performance probably in 5-8 times but if you do need mapping from source to existing destination then it is valid time to compile 12MB of text. I'm going to create new ticket to compile only source mappings.

Looking forward for your feedback.

Thanks,
Yuriy

**Comment by Mariana1980** [comment]
Hi Yuriy, thanks for your reply.
With this WPF app, the first time the mapper delays 2 minutes, second time only few seconds.
My processor is Intel Core(TM)i5-3470 CPU@3.20HGz. 
RAM is 8 GB.
Can you test this application and tell us which are your test times.
You can downloaded from [here] (https://www.dropbox.com/s/zgks9uy8pdima3d/TestExpressMapper.rar?dl=0)
Thanks a lot.

**Comment by anisimovyuriy** [comment]
Dear Mariana1980,

It is fixed and it is included in 1.7 release.
Please use the following:
```Mapper.Compile(CompilationTypes.OnlySource)``` and ```Mapper.PreCompileCollection<SourceCollectionType, DestinationCollectionType>(CompilationTypes.OnlySource)```  - in case you would like to precompile or just use normal mapping without precompilation.

It is working for me lightlingly fast since these changes. Please let me know how it is working for you.
Thanks a lot for such great input!

Looking forward to hearing from you.

**Comment by Mariana1980** [comment]
Dear Yuriy!! 
We tested our application with 1.7 ExpressMapper and it works perfectly!!!!!!!! 
Thanks a lot for your soon answer!!!!!
We are in contact!!
